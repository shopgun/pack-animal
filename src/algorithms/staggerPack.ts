import { IPoint, polygonArea, polygonHeight, polygonWidth } from "../geometry";
import { standardDeviation } from "../maths";
import { getPolygonTransform, ITransform } from "../transform";
import { noop, numberRange, permutator } from "../utilities";
import { Matrix } from "../vendor/matrix";

export const staggerPack = (
  rectangleWidth: number,
  rectangleHeight: number,
  polygons: IPoint[][],
  { debug = noop } = {}
): ITransform[] => {
  // find desired grid dimensions
  /*
Given cellCount and gridRatio, find gridWidth and gridHeight such that:
gridWidth * gridHeight ≥ cellCount
gridWidth / gridHeight ≈ gridRatio
  */
  const cellCount = polygons.length;
  const rectangleRatio = rectangleWidth / rectangleHeight;
  debug({ rectangleRatio });
  const permutations = permutator(numberRange(1, cellCount), 2);

  const averageArea =
    polygons.reduce((memo, points) => memo + polygonArea(points), 0) / polygons.length;

  const normalizedScaleFromPoints = (points: IPoint[]) =>
    (averageArea / polygonArea(points) - 1) / 2 + 1;
  const maxWidth = Math.max(
    ...polygons.map(points => polygonWidth(points) * normalizedScaleFromPoints(points))
  );
  const maxHeight = Math.max(
    ...polygons.map(points => polygonHeight(points) * normalizedScaleFromPoints(points))
  );
  const minWidth = Math.min(
    ...polygons.map(points => polygonWidth(points) * normalizedScaleFromPoints(points))
  );
  const minHeight = Math.min(
    ...polygons.map(points => polygonHeight(points) * normalizedScaleFromPoints(points))
  );
  const averageRatio =
    polygons.reduce((memo, points) => memo + polygonWidth(points) / polygonHeight(points), 0) /
    polygons.length;
  const averageRatioInv = 1 + (1 - averageRatio);
  const averageRatios = [averageRatio, averageRatioInv].sort();

  const gridDimensionsToRatio = (w: number, h: number) =>
    maxWidth * w / (maxHeight + maxHeight / 2 * h);

  const [gridWidth, gridHeight] = permutations.reduce((bestPermutation, currentPermutation) => {
    const count = currentPermutation[0] * currentPermutation[1];
    const ratio = gridDimensionsToRatio(currentPermutation[0], currentPermutation[1]);
    debug({ w: currentPermutation[0], h: currentPermutation[1], ratio });
    if (count < cellCount) {
      debug("discarded small grid");
      return bestPermutation;
    }
    if (count - cellCount >= currentPermutation[0]) {
      debug("discarded empty row grid");
      return bestPermutation;
    }
    if (
      count > 4 &&
      currentPermutation[0] > 2 &&
      count !== cellCount &&
      currentPermutation[0] !== count &&
      currentPermutation[1] !== count &&
      count - cellCount === currentPermutation[0] - 1
    ) {
      debug("discard grid that would have an orphan");
      return bestPermutation;
    }
    // compare previous(best) permutations if present
    if (bestPermutation[0] && bestPermutation[1]) {
      const bestRatio = gridDimensionsToRatio(bestPermutation[0], bestPermutation[1]);
      return Math.abs(ratio - rectangleRatio) < Math.abs(bestRatio - rectangleRatio)
        ? currentPermutation
        : bestPermutation;
    }
    return currentPermutation;
  }, []);
  debug({ gridHeight, gridWidth });
  const isVerticalGrid = gridHeight > gridWidth;
  debug(standardDeviation([rectangleRatio, averageRatio]));
  const twoferPack =
    polygons.length === 2 && standardDeviation([rectangleRatio, averageRatio]) < 0.25;
  debug({ twoferPack });
  // TODO: if we're at this point, we have a 1xY or Xx1 grid and are not twoferpacking
  // We should use a separate packing algorithm specialized for these scenarios, as grids
  // can look quite stiff or unnaturally spaced when presented with only a single row or column.
  const horizontalOffset = maxWidth;
  const verticalOffset = maxHeight - (!isVerticalGrid ? maxHeight / 2 : 0);
  debug({
    horizontalOffset,
    verticalOffset
  });
  const polygonTransforms = polygons.map((polygon, index) => {
    const normalizingScale = normalizedScaleFromPoints(polygon);
    const x = index % gridWidth;
    const y = Math.floor(index / gridWidth);
    const horizontalStagger =
      !isVerticalGrid || twoferPack
        ? +!!(y % 2) * (maxWidth * (0.5 * (twoferPack ? averageRatios[0] : 1)))
        : 0;
    const verticalStagger =
      isVerticalGrid || twoferPack
        ? +!!(x % 2) * (maxHeight * (0.5 * (twoferPack ? averageRatios[0] : 1)))
        : 0;
    debug({
      horizontalStagger,
      verticalStagger
    });
    const horizontalCorrection = maxWidth - polygonWidth(polygon) * normalizingScale / 2;
    const verticalCorrection = maxHeight - polygonHeight(polygon) * normalizingScale / 2;
    debug({
      horizontalCorrection,
      verticalCorrection
    });

    return getPolygonTransform(
      rectangleWidth,
      rectangleHeight,
      polygon,
      new Matrix()
        .translate(
          x * horizontalOffset +
            horizontalStagger +
            horizontalCorrection +
            -(twoferPack && index === 1 && !isVerticalGrid
              ? minWidth * (0.5 * averageRatios[0])
              : 0),
          y * verticalOffset +
            verticalStagger +
            verticalCorrection +
            -(twoferPack && index === 1 && isVerticalGrid
              ? minHeight * (0.5 * averageRatios[0])
              : 0)
        )
        .scaleU(normalizingScale)
    );
  });
  return polygonTransforms;
};
