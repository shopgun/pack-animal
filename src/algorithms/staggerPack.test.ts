import { RotationMode } from "./";
import { staggerPack } from "./staggerPack";

describe("staggerPack", () => {
  it("works if true is truthy", () => {
    expect(true).toBeTruthy();
  });
  it("gives nothing for nothingg", () => {
    const polygonTransforms = staggerPack(0, 0, []);
    expect(polygonTransforms).toMatchSnapshot();
  });
  it("works", () => {
    expect(
      staggerPack(800, 742, [
        [{ x: 1, y: 1 }, { x: 1, y: 366 }, { x: 798, y: 366 }, { x: 798, y: 1 }],
        [{ x: 1, y: 1 }, { x: 41, y: 390 }, { x: 773, y: 390 }, { x: 798, y: 1 }],
        [{ x: 1, y: 1 }, { x: 41, y: 390 }, { x: 773, y: 390 }, { x: 798, y: 1 }]
      ])
    ).toMatchSnapshot();
  });
});
