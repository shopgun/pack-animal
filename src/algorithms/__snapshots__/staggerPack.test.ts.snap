// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`staggerPack gives nothing for nothingg 1`] = `Array []`;

exports[`staggerPack works 1`] = `
Array [
  Object {
    "cssText": "
        position: absolute;
        top: 0;
        left: 0;
        transform-origin: 0 0;
        width: 108.21062413075855%;
        height: 45.964057925554165%;
        z-index: 401;
        transform: translateX(50%) translateY(0%) rotate(0deg)
      ",
    "matrix": "{
      \\"a\\":1.0074314065874885,
      \\"b\\":0,
      \\"c\\":0,
      \\"d\\":1.0074314065874885,
      \\"e\\":401.4614155251142,
      \\"f\\":0
    }",
    "points": Array [
      Object {
        "x": 402.4688469317017,
        "y": 1.0074314065874885,
      },
      Object {
        "x": 402.4688469317017,
        "y": 368.7198948110208,
      },
      Object {
        "x": 1205.3916779819301,
        "y": 368.7198948110208,
      },
      Object {
        "x": 1205.3916779819301,
        "y": 1.0074314065874885,
      },
    ],
    "rotate": -0,
    "scale": 1.0074314065874885,
    "svgTransform": "matrix(1.0074314065874885,0,0,1.0074314065874885,401.4614155251142,0)",
    "translateX": 401.4614155251142,
    "translateY": 0,
    "zIndex": 401,
  },
  Object {
    "cssText": "
        position: absolute;
        top: 0;
        left: 0;
        transform-origin: 0 0;
        width: 107.02199017575236%;
        height: 48.448264115979946%;
        z-index: 1141;
        transform: translateX(51.110644600286534%) translateY(94.87245573034598%) rotate(0deg)
      ",
    "matrix": "{
      \\"a\\":0.9963653288633407,
      \\"b\\":0,
      \\"c\\":0,
      \\"d\\":0.9963653288633407,
      \\"e\\":405.87124749818713,
      \\"f\\":367.7124634044333
    }",
    "points": Array [
      Object {
        "x": 406.8676128270505,
        "y": 368.7088287332967,
      },
      Object {
        "x": 446.7222259815841,
        "y": 756.2949416611361,
      },
      Object {
        "x": 1176.0616467095497,
        "y": 756.2949416611361,
      },
      Object {
        "x": 1200.9707799311332,
        "y": 368.7088287332967,
      },
    ],
    "rotate": -0,
    "scale": 0.9963653288633407,
    "svgTransform": "matrix(0.9963653288633407,0,0,0.9963653288633407,405.87124749818713,367.7124634044333)",
    "translateX": 405.87124749818713,
    "translateY": 367.7124634044333,
    "zIndex": 1141,
  },
  Object {
    "cssText": "
        position: absolute;
        top: 0;
        left: 0;
        transform-origin: 0 0;
        width: 107.02199017575236%;
        height: 48.448264115979946%;
        z-index: 1916;
        transform: translateX(51.110644600286534%) translateY(194.87245573034596%) rotate(0deg)
      ",
    "matrix": "{
      \\"a\\":0.9963653288633407,
      \\"b\\":0,
      \\"c\\":0,
      \\"d\\":0.9963653288633407,
      \\"e\\":405.87124749818713,
      \\"f\\":755.2985763322729
    }",
    "points": Array [
      Object {
        "x": 406.8676128270505,
        "y": 756.2949416611362,
      },
      Object {
        "x": 446.7222259815841,
        "y": 1143.8810545889758,
      },
      Object {
        "x": 1176.0616467095497,
        "y": 1143.8810545889758,
      },
      Object {
        "x": 1200.9707799311332,
        "y": 756.2949416611362,
      },
    ],
    "rotate": -0,
    "scale": 0.9963653288633407,
    "svgTransform": "matrix(0.9963653288633407,0,0,0.9963653288633407,405.87124749818713,755.2985763322729)",
    "translateX": 405.87124749818713,
    "translateY": 755.2985763322729,
    "zIndex": 1916,
  },
]
`;
